WordPressKitTests/Utilities/URLSessionHelperTests.swift

```diff
@@ -0,0 +1,130 @@
+import Foundation
+import XCTest
+import OHHTTPStubs
+
+@testable import WordPressKit
+
+class URLSessionHelperTests: XCTestCase {
+
+    override func tearDown() {
+        super.tearDown()
+        HTTPStubs.removeAllStubs()
+    }
+
+    func testConnectionError() async throws {
+        stub(condition: isPath("/hello")) { _ in
+            HTTPStubsResponse(error: URLError(.serverCertificateUntrusted))
+        }
+
+        let result = await URLSession.shared.perform(request: .init(url: URL(string: "https://wordpress.org/hello")!), errorType: TestError.self)
+        do {
+            _ = try result.get()
+            XCTFail("The above call should throw")
+        } catch let WordPressAPIError<TestError>.connection(error) {
+            XCTAssertEqual(error.code, URLError.Code.serverCertificateUntrusted)
+        } catch {
+            XCTFail("Unknown error: \(error)")
+        }
+    }
+
+    func test200() async throws {
+        stub(condition: isPath("/hello")) { _ in
+            HTTPStubsResponse(data: "success".data(using: .utf8)!, statusCode: 200, headers: nil)
+        }
+
+        let result = await URLSession.shared.perform(request: .init(url: URL(string: "https://wordpress.org/hello")!), errorType: TestError.self)
+
+        // The result is a successful result. This line should not throw
+        _ = try result.get()
+
+        let expectation = expectation(description: "API call returns a successful result")
+        _ = result
+            .assessStatusCode { result in
+                XCTAssertEqual(String(data: result.body, encoding: .utf8), "success")
+                expectation.fulfill()
+                return result
+            } failure: { _ in
+                // Do nothing
+                return nil
+            }
+        await fulfillment(of: [expectation])
+    }
+
+    func testUnacceptable500() async throws {
+        stub(condition: isPath("/hello")) { _ in
+            HTTPStubsResponse(data: "Internal server error".data(using: .utf8)!, statusCode: 500, headers: nil)
+        }
+
+        let result = await URLSession.shared.perform(request: .init(url: URL(string: "https://wordpress.org/hello")!), errorType: TestError.self)
+
+        // The result is a successful result. This line should not throw
+        _ = try result.get()
```

Comment:
I had the same thought too. On one hand, what's done here follows `URLSession`'s API, where it doesn't take HTTP semantic into account. On the other hand, I do want to make these helpers somewhat opinionated: locking them down to suit for our cases rather than making it too general like `URLSession`.

We can add an "acceptable status codes" argument to the `perform(request:)` function. That means we'll need to add another error case `WordPressAPIError.unaccpetableStatusCode(HTTPAPIResponse<Data>)`. The helper that deals with success and error parsing (parsing generic JSON to specific model, parsing http error response to a concrete error type) will need to change too.

I think these changes may potentially be implemented later on. These helpers are internal API after they, they can be easily changed without much affects to others.

At the moment, I only looked into the code in `WordPressOAuthClient.swift`, and the helpers are all created based on the existing implementation. That's why I haven't put too much though into supporting `Decodable`, which I think is very likely going to happen once I start looking into the existing WP.com REST API implementation.